ocaml/xapi/xapi_sr_operations.ml

```diff
@@ -221,21 +224,21 @@ let valid_operations ~__context ?op record _ref' : table =
   table
 
 let throw_error (table : table) op =
-  if not (Hashtbl.mem table op) then
-    raise
-      (Api_errors.Server_error
-         ( Api_errors.internal_error
-         , [
-             Printf.sprintf
-               "xapi_sr.assert_operation_valid unknown operation: %s"
-               (sr_operation_to_string op)
-           ]
-         )
-      ) ;
-  match Hashtbl.find table op with
-  | Some (code, params) ->
-      raise (Api_errors.Server_error (code, params))
+  match Hashtbl.find_opt table op with
   | None ->
+      raise
+        (Api_errors.Server_error
```

Comment:
Define `internal_error fmt` in helpers and use that.
```
let fail fmt =
    Printf.kprintf
      (fun msg -> raise Api_errors.(Server_error (import_error_generic, [msg])))
      fmt
```

The above code is not raising `internal_error` but demonstrates how to define a function that raises an exception and takes printf-style arguments. You can use that to simplify other places that raise internal errors.

Reply:
Yeah, there is a similar function defined in `xenops_server.ml`, for example:

```
let internal_error fmt =
  Printf.kprintf
    (fun str ->
      error "%s" str ;
      raise (Xenopsd_error (Internal_error str))
    )
    fmt
```

Will open a separate ticket for this --- there are quite a few such cases with internal `Printf.sprintf`